File: src/main/java/com/herocc/school/aspencheck/aspen/course/AspenCoursesController.java
package com.herocc.school.aspencheck.aspen.course;

import com.herocc.school.aspencheck.AspenCheck;
import com.herocc.school.aspencheck.District;
import com.herocc.school.aspencheck.ErrorInfo;
import com.herocc.school.aspencheck.JSONReturn;
import com.herocc.school.aspencheck.aspen.AspenWebFetch;
import org.jsoup.Connection;
import org.jsoup.nodes.Element;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

@CrossOrigin
@RestController
@RequestMapping("/{district-id}/aspen")
public class AspenCoursesController {

  @RequestMapping("/course")
  public ResponseEntity<JSONReturn> serveSchedule(@PathVariable(value = "district-id") String districtName,
                                                  @RequestParam(value = "moreData", defaultValue = "false") String moreData,
                                                  @RequestHeader(value = "ASPEN_UNAME", required = false) String u,
                                                  @RequestHeader(value = "ASPEN_PASS", required = false) String p,
                                                  @RequestParam(value = "term", required = false) String term) {

    if (u != null && p != null) {
      return new ResponseEntity<>(new JSONReturn(getCourses(new AspenWebFetch(districtName, u, p), moreData.equals("true"), term), new ErrorInfo()), HttpStatus.OK);
    } else {
      return new ResponseEntity<>(new JSONReturn(null, new ErrorInfo("Invalid Credentials", 0, "No username or password given")), HttpStatus.UNAUTHORIZED);
    }
  }

  @RequestMapping("/course/{course-id}")
  public ResponseEntity<JSONReturn> serveCourseInfo(@PathVariable(value = "district-id") String districtName,
                                                    @PathVariable(value = "course-id") String course,
                                                    @RequestHeader(value = "ASPEN_UNAME", required = false) String u,
                                                    @RequestHeader(value = "ASPEN_PASS", required = false) String p,
                                                    @RequestParam(value = "term", required = false) String term) {

    if (u != null && p != null) {
      AspenWebFetch a = new AspenWebFetch(districtName, u, p);
      Course c = getCourse(a, course, term).getMoreInformation(a, term);
      if (c == null)
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(new JSONReturn(null, new ErrorInfo("Course not Found", 404, "The course you tried to fetch doesn't exist or was inaccessible")));
      return new ResponseEntity<>(new JSONReturn(c, new ErrorInfo()), HttpStatus.OK);
    } else {
      return new ResponseEntity<>(new JSONReturn(null, new ErrorInfo("Invalid Credentials", 0, "No username or password given")), HttpStatus.UNAUTHORIZED);
    }
  }

  public static List<Course> getCourses(AspenWebFetch a) { return getCourses(a, false); }

  public static List<Course> getCourses(AspenWebFetch a, boolean moreData) { return getCourses(a, moreData, null); }

  public static List<Course> getCourses(AspenWebFetch a, boolean moreData, String term) {
    AspenCheck.log.info("getting courses for " + a.districtName + ", term = " + term);
    Connection.Response classListPage;
    if(Arrays.asList("1", "2", "3", "4").contains(term)){
      classListPage = a.getCourseListPage(Integer.parseInt(term));
    }else {
      classListPage = a.getCourseListPage();
    }
    List<Course> courses = new ArrayList<>();
    if (classListPage != null) {
      try {
        for (Element classRow : classListPage.parse().body().getElementsByAttributeValueContaining("class", "listCell listRowHeight")) {
          Map<String, Integer> columnOrganization = AspenCheck.config.districts.containsKey(a.districtName) ? AspenCheck.config.districts.get(a.districtName).columnOrganization : District.defaultColumnOrganization();
          Course c = new Course(classRow, columnOrganization);
          c = moreData ? c.getMoreInformation(a, term) : c;
          courses.add(c);
        }
      } catch (IOException e) {
        e.printStackTrace();
        AspenCheck.rollbar.error(e, "Error while parsing CourseList of user from " + a.districtName);
      }
    }
    return courses;
  }

  public static Course getCourse(AspenWebFetch a, String courseId, boolean moreData, String term) {
    List<Course> enrolledCourses = getCourses(a);

    for (Course c : enrolledCourses) {
      if (c.id.equalsIgnoreCase(courseId) || c.code.equalsIgnoreCase(courseId) || c.name.equalsIgnoreCase(courseId))
        return moreData ? c.getMoreInformation(a, term) : c;
    }
    return null;
  }

  public static Course getCourse(AspenWebFetch a, String courseId, String term) {
    return getCourse(a, courseId, true, term);
  }
}

File: src/main/java/com/herocc/school/aspencheck/aspen/course/AspenCoursesController.kt
package com.herocc.school.aspencheck.aspen.course

import com.herocc.school.aspencheck.AspenCheck
import com.herocc.school.aspencheck.District
import com.herocc.school.aspencheck.ErrorInfo
import com.herocc.school.aspencheck.JSONReturn
import com.herocc.school.aspencheck.aspen.AspenWebFetch
import com.herocc.school.aspencheck.aspen.course.assignment.Assignment
import com.herocc.school.aspencheck.aspen.course.assignment.AssignmentInfoFetcher
import kotlinx.coroutines.*
import org.jsoup.Connection
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.scheduling.annotation.Async
import org.springframework.scheduling.annotation.EnableAsync
import org.springframework.web.bind.annotation.*
import java.io.IOException
import java.util.*
import java.util.concurrent.CompletableFuture

@CrossOrigin
@RestController
@RequestMapping("/{district-id}/aspen")
open class AspenCoursesController {

    @Async
    @RequestMapping("/course")
    fun serveCourseList(
        @PathVariable(value = "district-id") districtName: String,
        @RequestParam(value = "moreData", defaultValue = "false") moreData: String,
        @RequestHeader(value = "ASPEN_UNAME", required = false) u: String?,
        @RequestHeader(value = "ASPEN_PASS", required = false) p: String?,
        @RequestParam(value = "term", required = false) term: String?
    ): CompletableFuture<ResponseEntity<JSONReturn>> {
        return if (u != null && p != null) {
            val coursesManager = AspenCoursesManager(u, p, AspenCheck.getDistrictByName(districtName))
            if (!coursesManager.aspenWebFetch.areCredsCorrect()) {
                AspenCheck.incorrectCredentialsResponse()
            } else {
                val courses = coursesManager.getCourses(moreData.toLowerCase()=="true", term)
                if (moreData=="true") {
                    AspenCheck.log.info("all moredata gotten = ${courses.all { it.assignments!=null }}")
                }
                CompletableFuture.completedFuture(ResponseEntity(JSONReturn(courses, ErrorInfo()), HttpStatus.OK))
            }
        } else {
            AspenCheck.invalidCredentialsResponse()
        }
    }

    @Async
    @RequestMapping("/course/{course-id}")
    fun serveCourseInfo(
        @PathVariable(value = "district-id") districtName: String?,
        @PathVariable(value = "course-id") courseID: String?,
        @RequestHeader(value = "ASPEN_UNAME", required = false) u: String?,
        @RequestHeader(value = "ASPEN_PASS", required = false) p: String?,
        @RequestParam(value = "term", required = false) term: String?
    ): CompletableFuture<ResponseEntity<JSONReturn>> {
        return if (u!=null && p!=null) {
            val coursesManager = AspenCoursesManager(u, p, AspenCheck.getDistrictByName(districtName))
            if (!coursesManager.aspenWebFetch.areCredsCorrect()) {
                return AspenCheck.incorrectCredentialsResponse()
            }
            val course = coursesManager.getCourses(true, term).find { it.id==courseID }
                ?: return AspenCheck.incorrectCourseIDResponse()
            CompletableFuture.completedFuture(ResponseEntity(JSONReturn(course, ErrorInfo()), HttpStatus.OK))
        } else {
            AspenCheck.invalidCredentialsResponse()
        }
    }
}

class AspenCoursesManager(
    private val username: String,
    private val password: String,
    private val district: District,
    val aspenWebFetch: AspenWebFetch = AspenWebFetch(district.districtName, username, password)
) {

    fun getCourses(moreData: Boolean = false, term: String? = null): List<Course> {
        AspenCheck.log.info("getting courses for " + district.districtName + ", term = " + term)
        val classListPage: Connection.Response = when (term) {
            in listOf("1", "2", "3", "4") -> aspenWebFetch.getCourseListPage(term!!.toInt())
            else -> aspenWebFetch.getCourseListPage()
        }
        try {
            val courseElements = classListPage.parse().body().getElementsByAttributeValueContaining("class", "listCell listRowHeight")
            if(courseElements[0].getElementsContainingOwnText("No matching records").size > 0){
                return getCourses(moreData, "4")
            }
            val courses = courseElements.map { classRow ->
                Course(classRow, district.columnOrganization)
            }
            if (moreData) {
                getMoreInfoCourses(courses, term)
            }
            return courses
        } catch (e: IOException) {
            e.printStackTrace()
            AspenCheck.rollbar.error(e, "Error while parsing CourseList of user from " + district.districtName)
            return emptyList()
        }
    }

    fun getMoreInfoCourses(courses: List<Course>, term: String?) {
        runBlocking {
            val webFetches = (0 until 4).map { AspenWebFetch(district.districtName, username, password) }
            val jobs = courses.chunked(courses.size / webFetches.size).mapIndexed { i, coursesPerWebFetch ->
                CoroutineScope(Dispatchers.IO).launch {
                    val webFetch = webFetches[i]
                    if(term != null){
                        webFetch.getCourseListPage(term.toInt())
                    }
                    coursesPerWebFetch.forEach { course ->
                        //AspenCheck.log.info("webFetch for ${course.name} login = ${webFetch.areCredsCorrect()}")
                        course.getMoreInformation(webFetch, term)
                        //AspenCheck.log.info("webFetch after for ${course.name} login = ${webFetch.areCredsCorrect()}")
                    }
                }
            }
            jobs.joinAll()
        }
    }
}

fun getAssignmentList(a: AspenWebFetch, course: Course, term: String?): List<Assignment> {
    val assignmentsPage: Connection.Response = when (term) {
        in listOf("1", "2", "3", "4") -> a.getCourseAssignmentsPage(course.id, term!!.toInt())
        else -> a.getCourseAssignmentsPage(course.id)
    } ?: return emptyList()

    try {
        val assignmentElements = assignmentsPage.parse().body().getElementsByAttributeValueContaining("class", "listCell listRowHeight")
        val assignments = assignmentElements.map { assignmentRow ->
            if (assignmentRow.text().contains("No matching records")) return@map null
            Assignment(assignmentRow, AspenCheck.config.districts[a.districtName]!!.gradeScale)
        }.filterNotNull()

        AssignmentInfoFetcher().getAdditionalInfoForAssignments(a, course.id, assignments)
        return assignments

    } catch (e: IOException) {
        e.printStackTrace()
        return emptyList()
    }
}